62.CAD/fragment_shader.hlsl

```diff
@@ -15,6 +16,68 @@ void endInvocationInterlockEXT();
 // Write a general one, and maybe another one that uses precomputed values, and move these to somewhere nice in our builtin hlsl shaders if we don't have one
 // See: https://github.com/erich666/GraphicsGems/blob/master/gems/Roots3And4.c
 
+template<typename float_t, typename CurveType>
+struct LineStyleClipper
+{
+    using float2_t = vector<float_t, 2>;
+
+    static LineStyleClipper<float_t, CurveType> construct(uint32_t styleIdx, 
+                                                          CurveType curve,
+                                                          typename CurveType::ArcLenCalculator arcLenCalc,
+                                                          float totalArcLen)
+    {
+        LineStyleClipper<float_t, CurveType> ret = { styleIdx, curve, arcLenCalc, totalArcLen };
+        return ret;
+    }
+    
+    struct ArrayAccessor
+    {
+        uint32_t styleIdx;
+        using value_type = float;
+
+        float operator[](const uint32_t ix)
+        {
+            return lineStyles[styleIdx].stipplePattern[ix];
+        }
+    };
+    
+    float2_t operator()(const float t)
+    {
+        const float arcLen = arcLenCalc.calcArcLen(t);
+        float_t tMappedToPattern = frac(arcLen / float(globals.screenToWorldRatio) * lineStyles[styleIdx].recpiprocalStipplePatternLen + lineStyles[styleIdx].phaseShift);
+        ArrayAccessor stippleAccessor = { styleIdx };
+        uint32_t patternIdx = nbl::hlsl::upper_bound(stippleAccessor, 0, lineStyles[styleIdx].stipplePatternSize, tMappedToPattern);
+        
+        if(patternIdx & 0x1)
+        {   
+            float t0NormalizedLen = (patternIdx == 0) ? 1.0 : lineStyles[styleIdx].stipplePattern[patternIdx-1];
+            float t1NormalizedLen = (patternIdx == lineStyles[styleIdx].stipplePatternSize) ? 1.0 : lineStyles[styleIdx].stipplePattern[patternIdx];
+            t0NormalizedLen -= tMappedToPattern;
+            t1NormalizedLen -= tMappedToPattern;
+            
+            float t0 = t0NormalizedLen / globals.worldToScreenRatio / lineStyles[styleIdx].recpiprocalStipplePatternLen;
+            float t1 = t1NormalizedLen / globals.worldToScreenRatio / lineStyles[styleIdx].recpiprocalStipplePatternLen;
+            
+            t0 = arcLenCalc.calcArcLenInverse(curve, arcLen + t0, 0.000001f, arcLen / totalArcLen);
+            t1 = arcLenCalc.calcArcLenInverse(curve, arcLen + t1, 0.000001f, arcLen / totalArcLen);

```

Comment:
The t value gave you arcLen, what gives you arcLen+smallOffset?
So you could just pass t here instead of arcLen/totalArcLen xD
And no need to calculate total arclen anymore AND it's a good hint/start for our iterative algorithm

Reply:
Btw if you really wanted to do this correctly (arcLen+offset)/totalArcLen would be better guess. But not worth it to calculate total arcLen for each vertex and pass it here

Reply:
I can't find any good reason to even keep this or do things like (arcLen+offset)/totalArcLen. The t value you already have is the closest thing you can have as hint

Reply:
> So you could just pass t here instead of arcLen/totalArcLen xD

EXACTLY!

Reply:
however note that the `t` we pass into the function is unclamped (closest point found by SDF, can already be outside of `[0,1]`) so it needs to be clamped right after this line, also not const anymore
```cpp
const float arcLen = arcLenCalc.calcArcLen(t);
```

Reply:
Also one more thing to note, because of the SDF closest points being outside of `[0,1]` its possible for the `arcLen+t0` or `arcLen+t1` to become negative or greater than `arcLen`

These cases make the point invalid (off-bezier) and you can just set the `t0` or `t1` to `0.f` or `arcLen` as necessary (otherwise your Newton Raphson will do the full 32 iterations as it will never be able to get the difference within bounds)

Reply:
done